src/toil/server/cli/wes_cwl_runner.py

```diff
@@ -0,0 +1,448 @@
+
+from werkzeug.utils import secure_filename
+from toil.wdl.utils import get_version as get_wdl_version
+from wes_client.util import WESClient, wes_reponse as wes_response  # type: ignore
+
+
+"""
+A CWL runner that submits a workflow to a WES server, waits for it to finish,
+and outputs the results.
+
+
+Environment variables:
++----------------------------------+----------------------------------------------------+
+| TOIL_WES_ENDPOINT                | URL to the WES server to use for this WES-based    |
+|                                  | CWL runner.                                        |
++----------------------------------+----------------------------------------------------+
+| TOIL_WES_USER                    | Username to use with HTTP Basic Authentication to  |
+|                                  | log into the WES server.                           |
++----------------------------------+----------------------------------------------------+
+| TOIL_WES_PASSWORD                | Password to use with HTTP Basic Authentication to  |
+|                                  | log into the WES server.                           |
++----------------------------------+----------------------------------------------------+
+
+
+Example usage with cwltest:
+
+```
+cwltest --verbose \
+    --tool=toil-wes-cwl-runner \
+    --test=src/toil/test/cwl/spec_v12/conformance_tests.yaml \
+    -n=1-50 \
+    --timeout=2400 \
+    -j2 \
+    -- \
+    --wes_endpoint=http://localhost:8080 \
+    --disableCaching \
+    --clean=always \
+    --logDebug
+```
+"""
+
+logger = logging.getLogger(__name__)
+
+
+def generate_attachment_path_name(base_dir: str, file_path: str) -> str:
+    """
+    Generate a relative path under base_dir as input fields or as attachment
+    keys.
+
+    This is to detect attachment path names that are above the CWL workflow file
+    and modify the path names in both the File or Directory objects in
+    workflow_params and the attachment key.
+
+    For example, for the following CWL workflow where "hello.yaml" references
+    a file "message.txt",
+
+        ~/toil/workflows/hello.cwl
+        ~/toil/input_files/hello.yaml
+        ~/toil/input_files/message.txt
+
+    This may be run with the command:
+        toil-wes-cwl-runner hello.cwl ../input_files/hello.yaml
+
+    Where "message.txt" is resolved to "../input_files/message.txt".
+
+    For the workflow attachment "../input_files/message.txt", we must replace
+    this relative path to something else that is not above the CWL workflow
+    file, otherwise the Toil WES server will drop all the traversals up to
+    contain attachments within its "execution" folder.
+
+    TODO: we need to find a way to call this function and replace the relative
+     paths referenced inside the CWL workflow directly.
+     e.g.: https://github.com/common-workflow-language/cwl-v1.2/blob/1.2.1_proposed/tests/iwd/iwd-fileobjs1.cwl#L7-L11
```

Comment:
Rather then replacing the relative paths in the workflow files, an approach might be to scan everything, find the highest-level directory traversed to, and then send everything as relative paths down from there.

So in this example, we'd see that `~/toil` is the common ancestor of the hierarchy we are working in, and we'd send everything as paths down from there, using that as the `base_dir` instead of wherever the CWL file lives. We'd send the workflow file as `workflows/hello.cwl`, and send the inputs as `input_files/hello.yaml` and `input_files/message.txt`. Then if the CWL file wants to look at `../input_files/message.txt` relative to itself, it can.